src/stdlib_sorting_radix_sort.f90

```diff
@@ -0,0 +1,466 @@
+        if (use_internal_buffer) then
+            deallocate (buffer)
+        end if
+    end subroutine sp_radix_sort
+
+    pure module subroutine int64_radix_sort(array, work, reverse)
+        integer(kind=int64), dimension(:), intent(inout) :: array
+        integer(kind=int64), dimension(:), intent(inout), target, optional :: work
+        logical, intent(in), optional :: reverse
+        integer(kind=int_size) :: i, N, start, middle, end
+        integer(kind=int64), dimension(:), pointer :: buffer
+        integer(kind=int64) :: item
+        logical :: use_internal_buffer
+        N = size(array, kind=int_size)
+        if (present(work)) then
+            if (size(work, kind=int_size) < N) then
+                error stop "int64_radix_sort: work array is too small."
+            end if
+            use_internal_buffer = .false.
+            buffer => work
+        else
+            use_internal_buffer = .true.
+            allocate (buffer(N))
+        end if
+        call radix_sort_u64_helper(N, array, buffer)
+        if (array(1) >= 0 .and. array(N) < 0) then
+            start = 1
+            end = N
+            middle = (1 + N)/2
+            do while (.true.)
+                if (array(middle) >= 0) then
+                    start = middle + 1
+                else
+                    end = middle
+                end if
+                middle = (start + end)/2
+                if (start == end) exit
+            end do
+            buffer(1:(N - middle + 1)) = array(middle:N)
+            buffer(N - middle + 2:N) = array(1:middle - 1)
+            array(:) = buffer(:)
+        end if
+        if (optval(reverse, .false.)) then
+            do i = 1, N/2
+                item = array(i)
+                array(i) = array(N - i + 1)
+                array(N - i + 1) = item
+            end do
+        end if
+        if (use_internal_buffer) then
+            deallocate (buffer)
+        end if
+    end subroutine int64_radix_sort
+
+    module subroutine dp_radix_sort(array, work, reverse)
+        use iso_c_binding
+        real(kind=dp), dimension(:), intent(inout), target :: array
+        real(kind=dp), dimension(:), intent(inout), target, optional :: work
+        logical, intent(in), optional :: reverse
+        integer(kind=int_size) :: i, N, pos, rev_pos
+        integer(kind=int64), dimension(:), pointer :: arri64
+        integer(kind=int64), dimension(:), pointer :: buffer
+        real(kind=dp) :: item
+        logical :: use_internal_buffer
+        N = size(array, kind=int_size)
+        if (present(work)) then
+            if (size(work, kind=int_size) < N) then
+                error stop "sp_radix_sort: work array is too small."
+            end if
+            use_internal_buffer = .false.
+            call c_f_pointer(c_loc(work), buffer, [N])
+        else
+            use_internal_buffer = .true.
+            allocate (buffer(N))
+        end if
+        call c_f_pointer(c_loc(array), arri64, [N])
```

Comment:
Would the intrinsic `transfer` not be appropriate for this?
It sounds like a new way of `equivalence`.

Reply:
> Would the intrinsic `transfer` not be appropriate for this? It sounds like a new way of `equivalence`.

It needs `arri64` and `array` pointer to the same memory, it seems that `transfer` cannot achieve such operation. `equivalence` is not work for pointer array.